packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart

```diff
@@ -202,120 +202,235 @@ class HotRestartButton extends StatelessWidget {
-
   @override
   _RegisteredServiceExtensionButtonState createState() =>
       _RegisteredServiceExtensionButtonState();
 }
 
 class _RegisteredServiceExtensionButtonState
-    extends State<_RegisteredServiceExtensionButton> {
-  StreamSubscription<bool> _subscription;
-
-  bool _disabled = false;
+    extends State<_RegisteredServiceExtensionButton>
+    with _ServiceExtensionMixin, AutoDisposeBase, AutoDisposeMixin {
   bool _hidden = true;
 
   @override
   void initState() {
     super.initState();
     // Only show the button if the device supports the given service.
-    _subscription = serviceManager.hasRegisteredService(
-      widget.serviceDescription.service,
-      (registered) {
+    autoDispose(
+      serviceManager.hasRegisteredService(
+        widget.serviceDescription.service,
+        (registered) {
+          setState(() {
+            _hidden = !registered;
+          });
+        },
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (_hidden) return const SizedBox();
+
+    return InkWell(
+      onTap: () => invokeAndCatchErrors(widget.action),
+      child: Container(
+        constraints: const BoxConstraints.tightFor(width: 48.0, height: 48.0),
+        alignment: Alignment.center,
+        // TODO(djshuckerow): Just make these icons the right size to fit this box.
+        // The current size is a little tiny by comparison to our other
+        // material icons.
+        child: getIconWidget(widget.serviceDescription.icon),
+      ),
+    );
+  }
+}
+
+/// Control that toggles the value of [structuredErrors].
+class StructuredErrorsToggle extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return _ServiceExtensionSwitch(
+      service: structuredErrors,
+      describeError: (error) =>
+          'Failed to update structuredError settings: $error',
+    );
+  }
+}
+
+class _ServiceExtensionSwitch extends _ServiceExtensionWidget {
+  const _ServiceExtensionSwitch({
+    Key key,
+    this.service,
+    @required String Function(dynamic) describeError,
+  }) : super(
+          key: key,
+          // Don't show messages on success or when this toggle is in progress.
+          completedText: null,
+          inProgressText: null,
+          describeError: describeError,
+        );
+  final ToggleableServiceExtensionDescription service;
+  @override
+  _ServiceExtensionMixin<_ServiceExtensionWidget> createState() =>
+      _ServiceExtensionCheckboxState();
+}
+
+class _ServiceExtensionCheckboxState extends State<_ServiceExtensionSwitch>
+    with _ServiceExtensionMixin, AutoDisposeBase, AutoDisposeMixin {
+  bool checked = false;
+  @override
+  void initState() {
+    super.initState();
+    autoDispose(serviceManager.serviceExtensionManager.getServiceExtensionState(
+      widget.service.extension,
+      (data) {
         setState(() {
-          _hidden = !registered;
+          checked = data.value == widget.service.enabledValue;
```

Comment:
checked may not be the best variable name now. Perhaps replace with `enabled`?

Reply:
I'll use `value`.  The Switch and Checkbox widgets both use that name for their selection status.